api/src/main/java/javax/mvc/locale/LocaleResolver.java

```diff
@@ -0,0 +1,103 @@
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package javax.mvc.locale;
+
+import javax.mvc.MvcContext;
+import java.util.Locale;
+
+/**
+ * <p>Locale resolvers are used to determine the locale of the current request and are discovered
+ * using CDI.</p>
+ *
+ * <p>The MVC implementation is required to resolve the locale for each request following this
+ * algorithm:</p>
+ *
+ * <ol>
+ * <li>Gather the set of all implementations of this interface available for injection via
+ * CDI.</li>
+ * <li>Sort the set of implementations using priorities in descending order. Locale resolvers
+ * can be decorated with {@link javax.annotation.Priority} to indicate their priority. If no
+ * priority is explicitly defined, the priority is assumed to be <code>1000</code>.</li>
+ * <li>Call the method {@link #resolveLocale(LocaleResolverContext)}. If the resolver returns
+ * a valid locale, use this locale as the request locale. If the resolver returns
+ * <code>null</code>, proceed with the next resolver in the ordered set.</li>
+ * </ol>
+ *
+ * <p>Controllers, view engines and other components can access the resolved locale by calling
+ * {@link MvcContext#getLocale()}.</p>
+ *
+ * <p>The MVC implementation is required to provide a default locale resolver with a priority
+ * of <code>0</code> which uses the <code>Accept-Language</code> request header to obtain the
+ * locale. If resolving the locale this way isn't possible, the implementation should return
+ * the application default locale specified via the {@link #DEFAULT_LOCALE} configuration
+ * property. If the application default locale hasn't been set, the resolver must return
+ * {@link Locale#getDefault()}.</p>
+ *
+ * @author Christian Kaltepoth
+ * @see javax.mvc.locale.LocaleResolverContext
+ * @see MvcContext#getLocale()
+ * @see java.util.Locale
+ * @since 1.0
+ */
+public interface LocaleResolver {
+
+    /**
+     * Name of the property that can be set in an application's {@link javax.ws.rs.core.Configuration}
+     * to set a application default locale. This locale will be used if the locale cannot be
+     * determined via the <code>Accept-Language</code> header sent by the client.
+     *
+     * @see javax.ws.rs.core.Application#getProperties()
+     */
+    String DEFAULT_LOCALE = "javax.mvc.locale.LocaleResolver.defaultLocale";
+
+    /**
+     * <p>Resolve the locale of the current request given a {@link LocaleResolverContext}.</p>
+     *
+     * <p>If the implementation is able to resolve the locale for the request, the corresponding
+     * locale must be returned. If the implementation cannot resolve the locale, it must return
+     * <code>null</code>. In this case the resolving process will continue with the next
+     * resolver.</p>
+     *
+     * @param context the context needed for processing.
+     * @return The resolved locale or <code>null</code>.
+     */
+    Locale resolveLocale(LocaleResolverContext context);
```

Comment:
> The implementation could still return null for some reason. The compiler won't help you here.

We should always return `Optional.ofNullable()` for nullable objects, so it will never be null, right?

our implementation could move from:
```
for (LocaleResolver candidate : candidates) {
      Locale locale = candidate.resolveLocale(context);
      if (locale != null) {
         return locale;
      }
}

throw new IllegalStateException("Could not resolve locale ...");
```
to a bit more functional style:
```
candidates.stream()
    .map(c -> c.resolveLocale(context))
    .findFirst()
    .orElseThrow(() -> new IllegalStateException("Could not resolve locale ..."));
}
```
I'm ok using null here, but IMHO we should use as many as possible the jdk8 possibilities